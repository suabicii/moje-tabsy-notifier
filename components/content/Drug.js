import React, {useState} from "react";
import {View} from "react-native";
import {Title} from "react-native-paper";
import DosingMoments from "./DosingMoments";
import {useDispatch, useSelector} from "react-redux";
import {List} from 'react-native-paper';
import {addExpandedAccordion, removeExpandedAccordion} from "../../features/detailsExpanded/expandedAccordionsSlice";

function Drug({drug}) {
    const {dosing, name, unit, dosingMoments} = drug;
    const drugsTaken = useSelector(state => state.drugsTaken);
    const expandedAccordions = useSelector(state => state.expandedAccordions);
    const dispatch = useDispatch();
    const [expanded, setExpanded] = useState(!!expandedAccordions.find(accordion => accordion.id === name));

    const handlePress = () => {
        if (!expanded) {
            dispatch(addExpandedAccordion({id: name}));
            setExpanded(true);
        } else {
            dispatch(removeExpandedAccordion(name));
            setExpanded(false);
        }
    };

    const dosingMomentsArray = Object.entries(dosingMoments);
    const borderStyles = {
        borderColor: '#dadada',
        borderRadius: 4,
        borderStyle: 'solid',
        borderWidth: 1.75,
    };
    let currentDrugsTaken = [];

    for (const [key] of dosingMomentsArray) {
        currentDrugsTaken = [
            ...currentDrugsTaken,
            ...drugsTaken.filter(string => string === `${name}_${key}`)
        ];
    }

    return (
        <>
            {
                // do not render drug component (content) if all doses was already taken
                currentDrugsTaken.length !== dosingMomentsArray.length
                &&
                <View style={{
                    ...borderStyles,
                    marginBottom: 10,
                    paddingBottom: 10,
                    paddingHorizontal: 15,
                }}>
                    <Title
                        testID={name}
                        style={{textAlign: "center"}}
                    >
                        {name} {dosing} {unit} w godz.
                    </Title>
                    <DosingMoments
                        drug={drug}
                        content={dosingMomentsArray}
                    />
                    <List.Section style={{...borderStyles}}>
                        <List.Accordion
                            testID={`details-${name}`}
                            title="Wszystkie godziny zaÅ¼ywania"
                            style={{backgroundColor: 'rgba(227,226,226,0.3)'}}
                            expanded={expanded}
                            theme={{ colors: { primary: '#000' }}}
                            onPress={handlePress}>
                            {Object.values(dosingMoments).map((hour, index) => (
                                <List.Item title={hour} titleStyle={{textAlign: "center"}} key={index}/>
                            ))}
                        </List.Accordion>
                    </List.Section>
                </View>
            }
        </>
    );
}

export default Drug;