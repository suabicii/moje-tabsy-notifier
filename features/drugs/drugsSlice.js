import {createSlice} from "@reduxjs/toolkit";
import {ajaxCall} from "../../utils/ajaxCall";


const drugsSlice = createSlice({
    name: 'drugs',
    initialState: [],
    reducers: {
        setDrugs: (state, action) => action.payload
    }
});

const {actions, reducer} = drugsSlice;

export const {setDrugs} = actions;

export const fetchDrugs = token => async dispatch => {
    await ajaxCall('get', `drug-notify/${token}`)
        .then(data => {
            dispatch(setDrugs(data));
        })
        .catch(err => {
            console.log(err);
        });
};

export default reducer;