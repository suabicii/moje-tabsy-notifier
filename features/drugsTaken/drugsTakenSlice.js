import {createSlice} from "@reduxjs/toolkit";

// State = array for storing strings associated
// with dosing moment keys -> if user clicks confirmation button, key will be pushed to the array, so thanks to it
// dosing moment won't be rendered again later

const drugsTakenSlice = createSlice({
    name: 'drugsTaken',
    initialState: [],
    reducers: {
        setDrugsTaken: (state, action) => action.payload
    }
});

const {actions, reducer} = drugsTakenSlice;

export const {setDrugsTaken} = actions;

export default reducer;