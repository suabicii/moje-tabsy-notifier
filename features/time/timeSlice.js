import {createSlice} from "@reduxjs/toolkit";

const timeSlice = createSlice({
    name: 'time',
    initialState: {},
    reducers: {
        setCurrentTime: (state, action) => ({...state, now: action.payload}),
        setTomorrowTime: (state, action) => ({...state, tomorrow: action.payload})
    }
});

const {actions, reducer} = timeSlice;

export const {setCurrentTime, setTomorrowTime} = actions;

export default reducer;