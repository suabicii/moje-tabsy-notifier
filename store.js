import {configureStore} from "@reduxjs/toolkit";
import drugsSlice from "./features/drugs/drugsSlice";
import timeSlice from "./features/time/timeSlice";
import drugsTakenSlice from "./features/drugsTaken/drugsTakenSlice";
import expandedAccordionsSlice from "./features/detailsExpanded/expandedAccordionsSlice";
import activeSessionSlice from "./features/activeSession/activeSessionSlice";

export default configureStore({
    reducer: {
        activeSession: activeSessionSlice,
        drugs: drugsSlice,
        drugsTaken: drugsTakenSlice,
        expandedAccordions: expandedAccordionsSlice,
        time: timeSlice
    }
});