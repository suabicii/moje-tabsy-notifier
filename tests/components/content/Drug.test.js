import React from "react";
import {Provider} from "react-redux";
import store from "../../../store";
import renderer from "react-test-renderer";
import Drug from "../../../components/content/Drug";
import {drugList} from "../../fixtures/drugList";
import MockDate from "mockdate";
import dayjs from "dayjs";
import {setCurrentTime} from "../../../features/time/timeSlice";
import {setDrugs} from "../../../features/drugs/drugsSlice";
import {act, fireEvent, render, screen, waitFor} from "@testing-library/react-native";
import {setDrugsTaken} from "../../../features/drugsTaken/drugsTakenSlice";

const drug = drugList[0];

beforeAll(() => {
    MockDate.set('2020-01-01');
    const currentTime = dayjs();
    store.dispatch(setCurrentTime(JSON.stringify(currentTime)));
    store.dispatch(setDrugs(drugList));
});

const dispatchDrugsTaken = () => {
    const dosingMoments = Object.entries(drug.dosingMoments);
    const drugsTaken = [];
    for (const [key] of dosingMoments) {
        drugsTaken.push(`${drug.name}_${key}`);
    }
    store.dispatch(setDrugsTaken(drugsTaken));
};

function WrappedComponent() {
    return (
        <Provider store={store}>
            <Drug drug={drugList[0]}/>
        </Provider>
    );
}

it('should correctly render Drug component', () => {
    const tree = renderer.create(<WrappedComponent/>).toJSON();
    expect(tree).toMatchSnapshot();
});

it('should not render single drug component if all doses was already taken', async () => {
    dispatchDrugsTaken();

    render(<WrappedComponent/>);

    await waitFor(() => {
        expect(screen.queryByTestId(drug.name)).toBeFalsy();
    });
});

it('should add id of accordion with drug details to global state after accordion expansion', async () => {
    render(<WrappedComponent/>);

    await act(() => {
        fireEvent.press(screen.getByTestId(`details-${drug.name}`));
    });

    expect(store.getState().expandedAccordions.length).toBeGreaterThan(0);
});

it('should remove id of accordion with drug details from global state after accordion narrowing', async () => {
    render(<WrappedComponent/>);

    await act(() => {
        fireEvent.press(screen.getByTestId(`details-${drug.name}`)); // expand accordion
    });
    await act(() => {
        fireEvent.press(screen.getByTestId(`details-${drug.name}`)); // return to previous state
    });

    expect(store.getState().expandedAccordions.length).toBeLessThan(1);
});