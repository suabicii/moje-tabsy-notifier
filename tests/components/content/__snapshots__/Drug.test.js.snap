// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Drug component 1`] = `
<View
  style={
    {
      "borderColor": "#dadada",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1.75,
      "marginBottom": 10,
      "paddingBottom": 10,
      "paddingHorizontal": 15,
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        {
          "textAlign": "left",
        },
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "System",
            "fontWeight": "500",
            "writingDirection": "ltr",
          },
          [
            {
              "fontSize": 20,
              "letterSpacing": 0.15,
              "lineHeight": 30,
              "marginVertical": 2,
            },
            {
              "textAlign": "center",
            },
          ],
        ],
      ]
    }
    testID="Xanax"
  >
    Xanax
     
    1
     
    szt.
     w godz.
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          {
            "textAlign": "left",
          },
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "400",
              "writingDirection": "ltr",
            },
            [
              {
                "fontSize": 14,
                "letterSpacing": 0.25,
                "lineHeight": 20,
                "marginVertical": 2,
              },
              {
                "fontSize": 16,
                "textAlign": "center",
              },
            ],
          ],
        ]
      }
    >
      – 
      07:00
    </Text>
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#6cc3d5",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "color": "#e8e8e8",
          "elevation": 0,
          "marginLeft": 5,
          "minWidth": "auto",
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "disabled": true,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 4,
            },
          ]
        }
        testID="Xanax_hour1"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              {
                "flexDirection": "row-reverse",
              },
            ]
          }
        >
          <View
            style={
              {
                "marginLeft": -4,
                "marginRight": 12,
              }
            }
          >
            <Text />
          </View>
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  {
                    "marginHorizontal": 8,
                  },
                  {
                    "textTransform": "uppercase",
                  },
                  {
                    "color": "rgba(0, 0, 0, 0.32)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Już zażyłem/-am
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          {
            "textAlign": "left",
          },
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "400",
              "writingDirection": "ltr",
            },
            [
              {
                "fontSize": 14,
                "letterSpacing": 0.25,
                "lineHeight": 20,
                "marginVertical": 2,
              },
              {
                "fontSize": 16,
                "textAlign": "center",
              },
            ],
          ],
        ]
      }
    >
      – 
      18:00
    </Text>
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#6cc3d5",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "color": "#e8e8e8",
          "elevation": 0,
          "marginLeft": 5,
          "minWidth": "auto",
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "disabled": true,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 4,
            },
          ]
        }
        testID="Xanax_hour2"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              {
                "flexDirection": "row-reverse",
              },
            ]
          }
        >
          <View
            style={
              {
                "marginLeft": -4,
                "marginRight": 12,
              }
            }
          >
            <Text />
          </View>
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  {
                    "marginHorizontal": 8,
                  },
                  {
                    "textTransform": "uppercase",
                  },
                  {
                    "color": "rgba(0, 0, 0, 0.32)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Już zażyłem/-am
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "marginVertical": 8,
        },
        {
          "borderColor": "#dadada",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 1.75,
        },
      ]
    }
    theme={
      {
        "animation": {
          "scale": 1,
        },
        "colors": {
          "accent": "#03dac4",
          "backdrop": "rgba(0, 0, 0, 0.5)",
          "background": "#f6f6f6",
          "disabled": "rgba(0, 0, 0, 0.26)",
          "error": "#B00020",
          "notification": "#f50057",
          "onSurface": "#000000",
          "placeholder": "rgba(0, 0, 0, 0.54)",
          "primary": "#6200ee",
          "surface": "#ffffff",
          "text": "#000000",
        },
        "dark": false,
        "fonts": {
          "light": {
            "fontFamily": "System",
            "fontWeight": "300",
          },
          "medium": {
            "fontFamily": "System",
            "fontWeight": "500",
          },
          "regular": {
            "fontFamily": "System",
            "fontWeight": "400",
          },
          "thin": {
            "fontFamily": "System",
            "fontWeight": "100",
          },
        },
        "roundness": 4,
      }
    }
  >
    <View>
      <View
        style={
          {
            "backgroundColor": "#f6f6f6",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "disabled": false,
              "expanded": false,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              [
                {
                  "padding": 8,
                },
                {
                  "backgroundColor": "rgba(227,226,226,0.3)",
                },
              ],
            ]
          }
          testID="details-Xanax"
        >
          <View
            pointerEvents="none"
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                [
                  {
                    "margin": 8,
                  },
                  {
                    "flex": 1,
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    {
                      "textAlign": "left",
                    },
                    [
                      {
                        "fontSize": 16,
                      },
                      {
                        "color": "rgba(0, 0, 0, 0.87)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Wszystkie godziny zażywania
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "margin": 8,
                  },
                  undefined,
                ]
              }
            >
              <Text />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
